utils/zerofrom/derive/src/lib.rs

```diff
@@ -0,0 +1,154 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! Custom derives for `ZeroFrom` from the `zerofrom` crate.
+
+use proc_macro::TokenStream;
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use quote::quote;
+use syn::spanned::Spanned;
+use syn::{parse_macro_input, parse_quote, DeriveInput, Ident, Lifetime, Type, WherePredicate};
+use synstructure::Structure;
+
+mod visitor;
+
+/// Custom derive for `zerofrom::ZeroFrom`,
+///
+/// This implements `ZeroFrom<Ty> for Ty` for types
+/// without a lifetime parameter, and `ZeroFrom<Ty<'data>> for Ty<'static>`
+/// for types with a lifetime parameter.
+///
+/// Apply the `#[zerofrom(cloning_zf)]` attribute if you wish for this custom derive
+/// to use `.clone()` for its implementation. The attribute can be applied to
+/// fields as well.
+#[proc_macro_derive(ZeroFrom, attributes(zerofrom))]
```

Comment:
I'm torn on the naming. On the one hand it makes sense to call the trait `ZeroFrom`, because it can convert from multiple types (unlike `Copy`/`Clone`), on the other hand it is weird to derive a `From`, as that needs a target type. This is deriving `ZeroFrom<Self>` which feels like more of a `ZeroClone`.